clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/CachedKafkaMetadataGetter.java

```diff
@@ -93,13 +97,21 @@ private <K, T> T fetchMetadata(K key, Map<K, ValueAndExpiryTime<T>> metadataCach
     if (cachedValue != null && cachedValue.getExpiryTimeNs() > now) {
       return cachedValue.getValue();
     }
-    T newValue = valueSupplier.get();
+
     if (cachedValue == null) {
+      T newValue = valueSupplier.get();
       cachedValue = new ValueAndExpiryTime<>(newValue, now + ttlNs);
       metadataCache.put(key, cachedValue);
-    } else {
-      // Reuse the existing object instead of creating a one to be more memory efficient.
-      cachedValue.setValueAndExpiryTimeMs(newValue, now + ttlNs);
+      return newValue;
+    }
+
+    // Update the value in the async fashion, for a given key and cache, we will only issue one async request at the
+    // same time.
+    if (cachedValue.valueUpdateInProgress.compareAndSet(false, true)) {
+      CompletableFuture.runAsync(() -> {
+        T newValue = valueSupplier.get();
```

Comment:
Can you put a `try/catch` around this line?
If there is any exception happened, we should mark `cacheValue.valueUpdateInProgress` to be `false`, otherwise, the cache for this entry will never be refreshed in the future.

And yes, it is good to add unit test to test different scenarios.